### PR TITLE
TOPO: compute all socket subgroups

### PR DESCRIPTION
**Pull request by vspetrov**
## What
1. Refactors TOPO framework:
     - moves topo related code (sbgp/topo) to separate location under components/topo
     - better naming for topo structs that represents what they are
     - documentation
     - change api to create a "team" topo: now it takes generic ucc_subset_t struct instead of team
2. Implements ucc_topo_get_all_sockets API. It is required for TL/SHM implementation. 
     - topo API, sbgp implementation + gtest

## Why ?
Refactoring is required for 2 reasons: we are in the process of development common topo framework that will be extended with additional functionality. It can potentially have some special components (e.g. cuda nvml for connectivity discovery) that will be loadable. We also want to keep all topo related stuff in one place.

The get_all_sockets API is required per TL/SHM design spec.

### CODE REVIEW

File: src/components/topo/ucc_sbgp.c [hunk 1]
```diff
@@ -20,16 +20,42 @@ const char*  ucc_sbgp_str(ucc_sbgp_type_t type)
...
+}
+
+
+static inline ucc_status_t sbgp_create_socket(ucc_topo_t *topo,
+                                              ucc_sbgp_t *sbgp,
+                                              ucc_rank_t  group_rank)
+{
+    ucc_sbgp_t   *node_sbgp  = &topo->sbgps[UCC_SBGP_NODE];
+    ucc_rank_t    nlr        = topo->node_leader_rank;
+    ucc_rank_t    sock_rank = 0, sock_size = 0;
```
Comment (shimmybalsam) [inline src/components/topo/ucc_sbgp.c:56]: alignment

File: src/components/topo/ucc_sbgp.h [hunk 2]
```diff
@@ -43,27 +43,41 @@ typedef enum ucc_sbgp_status_t {
...
 } ucc_sbgp_t;
 
 const char*  ucc_sbgp_str(ucc_sbgp_type_t type);
-ucc_status_t ucc_sbgp_create(ucc_team_topo_t *topo, ucc_sbgp_type_t type);
+
+/* The call creates a required subgroup specified by @in type in
+   the topo->sbgps[type]. The created sbgp can be in either of 3 states:
+   - NOT_EXISTS: means for a given topo (ucc team layout) there is no such
+     grouping or the result group is of size 1. Example: 1) type == SBGP_SOCKET
+     but processes are not bound to sockets at all; 2) type == SBGP_NEODE_LEADERS
```
Comment (shimmybalsam) [inline src/components/topo/ucc_sbgp.h:62]: SBGP_NODE_LEADERS (instead of NEODE)

File: src/components/topo/ucc_sbgp.c [hunk 3]
```diff
@@ -412,3 +435,51 @@ void ucc_sbgp_print(ucc_sbgp_t *sbgp)
...
+        if (UCC_OK != status) {
+            ucc_error("failed to create socket sbgp for sl_rank %d:%u", i, sl_rank);
+            goto error;
+        }
+        if (sbgps[i].rank_map) {
+            sbgps[i].map = ucc_ep_map_from_array(&sbgps[i].rank_map, sbgps[i].group_size,
+                                                 ucc_subset_size(&topo->set), 1);
+        }
+    }
+    *_sbgps = sbgps;
```
Comment (shimmybalsam) [inline src/components/topo/ucc_sbgp.c:479]: alignment

File: src/components/topo/ucc_sbgp.c [hunk 4]
```diff
@@ -74,17 +100,17 @@ static inline ucc_status_t sbgp_create_socket(ucc_team_topo_t *topo,
     return UCC_OK;
 }
 
-static inline ucc_status_t sbgp_create_node(ucc_team_topo_t *topo,
+static inline ucc_status_t sbgp_create_node(ucc_topo_t *topo,
                                             ucc_sbgp_t *     sbgp)
```
Comment (shimmybalsam) [inline src/components/topo/ucc_sbgp.c:104]: alignment

**Review by bureddy — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**